### PR TITLE
Uniformize semantics of :> in class after deprecation phase of #16230

### PR DESCRIPTION
**Pull request by proux01**
This ends the deprecation phases of https://github.com/coq/coq/pull/16230 and https://github.com/coq/coq/pull/15802 that introduced :: instead of :> for subclasses in Class declarations. Now :> produces an error in `Class` before we can make it mean coercion in all record declarations (including typeclasses) in next version.

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  
  - [x] Documented any new / changed **user messages**.
  - [ ] ~Updated **documented syntax** by running `make doc_gram_rsts`.~

#### Overlays (to be merged before the current PR):

To prepare overlays, just ensure that the code compiles with any Coq >= 8.18 and options `-w +future-coercion-class-constructor` and `-w +future-coercion-class-field`

* [x] https://github.com/math-comp/analysis/pull/1168
* [x] https://github.com/coq-community/atbr/pull/42
* [x] https://gitlab.inria.fr/coquelicot/coquelicot/-/commit/ca1a747aa8b7ccbfa67a55ae5c8e5c8df71cc396
* [x] https://github.com/coq-community/coq-ext-lib/pull/142
* [x] https://github.com/coq-community/corn/pull/203
* [x] https://github.com/DeepSpec/InteractionTrees/pull/261
* [x] https://github.com/coq-community/math-classes/pull/123
* [x] https://github.com/proux01/menhir/commit/60eaa1f64d3544a7bff95b97ac0dd08257a58595
* [x] https://github.com/QuickChick/QuickChick/pull/353
* [x] https://github.com/damien-pous/relation-algebra/pull/41
* [x] https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/537
* [x] https://github.com/MetaCoq/metacoq/pull/1055
* [x] https://github.com/mit-plv/bedrock2/pull/402
* [x] https://github.com/mit-plv/fiat-crypto/pull/1821
* [x] https://github.com/mit-plv/fiat/pull/105
* [x] https://github.com/LPCIC/coq-elpi/pull/607
* [x] https://github.com/smtcoq/smtcoq/pull/138
* [x] https://github.com/mit-plv/fiat-crypto/pull/1827 (fiat-crypto legacy)
* [x] https://github.com/mit-plv/fiat-crypto/pull/1830  (fiat-crypto legacy)

### CODE REVIEW

**Comment by proux01** [comment]
@fpottier @jhjourdan here is the patch for menhirlib (successfully tested in Coq's CI, including reverse dependencies compcert and json), backward compatible with very old Coq (at least 8.7 as currently required by menhir, a nicer solution exists but requires Coq >= 8.18)
```patch
From 60eaa1f64d3544a7bff95b97ac0dd08257a58595 Mon Sep 17 00:00:00 2001
From: Pierre Roux <pierre.roux@onera.fr>
Date: Wed, 31 Jan 2024 15:49:13 +0100
Subject: [PATCH] Adapt to https://github.com/coq/coq/pull/18590

---
 coq-menhirlib/src/Alphabet.v | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/coq-menhirlib/src/Alphabet.v b/coq-menhirlib/src/Alphabet.v
index beaa3268..3a19ce22 100644
--- a/coq-menhirlib/src/Alphabet.v
+++ b/coq-menhirlib/src/Alphabet.v
@@ -156,10 +156,13 @@ Class Finite (A:Type) := {
 
 (** An alphabet is both [ComparableLeibnizEq] and [Finite]. **)
 Class Alphabet (A:Type) := {
-  AlphabetComparable :> Comparable A;
-  AlphabetComparableLeibnizEq :> ComparableLeibnizEq AlphabetComparable;
-  AlphabetFinite :> Finite A
+  AlphabetComparable : Comparable A;
+  AlphabetComparableLeibnizEq : ComparableLeibnizEq AlphabetComparable;
+  AlphabetFinite : Finite A
 }.
+#[global] Existing Instance AlphabetComparable.
+#[global] Existing Instance AlphabetComparableLeibnizEq.
+#[global] Existing Instance AlphabetFinite.
 
 (** The [Numbered] class provides a conveniant way to build [Alphabet] instances,
    with a good computationnal complexity. It is mainly a injection from it to
-- 
2.34.1
```

**Comment by fpottier** [comment]
> here is the patch for menhirlib

Thanks! I have applied it. It will be in the next release of Menhir.

**Comment by proux01** [comment]
Thanks, no hurry you'll only need a release to support Coq 8.20 which shouldn't be released before summer.

**Comment by proux01** [comment]
So, the only remaining overlays are
* stdpp, iris and perennial (stdpp/iris in progress: https://gitlab.mpi-sws.org/iris/stdpp/-/merge_requests/537#note_100332 https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/1029 )
* bedrock2, fiat_crypto{,_ocaml,_legacy), fiat_parsers @samuelgruetter @andres-erbsen @JasonGross you just need to ensure that the code compiles with any Coq >= 8.18 and options `-w future-coercion-class-constructor` and `-w future-coercion-class-field` I may provide help but then I need to know which of the two solutions below you prefer:
```Coq
Class Foo := {
#[global] field :: type;
}.
```
or
```Coq
Class Foo := {
field : type;
}.
#[global] Existing Instance field.
```
(the former is nicer but requires Coq >= 8.18)

**Comment by JasonGross** [comment]
I have no strong preference personally and will happily accept a PR that does either (right now I am unfortunately too busy to prepare a PR myself, I may manage to get to it in a week or so.)  My very slight preferences are:
- For fiat_parsers and fiat_crypto_legacy let's go with the second one (means I don't need to fuss with the CI)
- For fiat_crypto, bedrock2, (and rupicola if it shows up) let's go with the first (we're dropping compact with Coq < 8.18 anyway)
- if this shows up on rewriter (I take it rewriter is fine though?), let's go with the first one (more backwards compatible)

Btw, it should be possible to avoid the dance of needing to wait for us to update submodules: if you set up the CI scripts so that when url is empty but, e.g., submodule_of_project is not, the scripts automatically clone the patent project and navigate to a specified subdirectory, and we set up separate CI developments coq-side for rupicola and bedrock2's dependencies much like bedrock2 is currently set up, and then pass along various `EXTERNAL_*=1` to `make`, submodule handling should be transparent to overlay writers.

File: doc/sphinx/addendum/type-classes.rst [hunk 1]
```diff
@@ -335,16 +335,16 @@ Command summary
 
 .. _warn-future-coercion-class-field:
 
-   .. warn:: A coercion will be introduced instead of an instance in future versions when using ':>' in 'Class' declarations. Replace ':>' with '::' (or use '#[global] Existing Instance field.' for compatibility with Coq < 8.17).
+   .. exn:: A coercion will be introduced instead of an instance in next version when using ':>' in 'Class' declarations. Replace ':>' with '::' (or use '#[global] Existing Instance field.' for compatibility with Coq < 8.17).
```
Comment (SkySkimmer) [inline doc/sphinx/addendum/type-classes.rst:338]: this is still a warning not an error even if the default produces an error

**Review by SkySkimmer — CHANGES_REQUESTED** [review]
I don't think we should change "future" to "next version".

Basically I think the warning shouldn't be changed (except for changing the default status) and the doc shouldn't be changed (except for the `..index:: :: (substructure)` bit which seems like it was an oversight).

**Comment by proux01** [comment]
> Basically I think the warning shouldn't be changed (except for changing the default status)

I do think we should also change the warning name. Otherwise, users who silenced the warning will miss the error, which would largely anihiliate the point of making the warning default to error, one version before actually performing the change.

**Comment by SkySkimmer** [comment]
My reasoning is the exact opposite, users who silenced the warning want to deal with it on their own and if we change the warning name they'll silence the new name too. So we would just be giving them useless work to do.

**Comment by proux01** [comment]
Ok, let's do that.

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
Missing overlay for fiat crypto legacy?

**Comment by proux01** [comment]
Ping @JasonGross (fiat crypto legacy is allow failure, I made almost all other overlays but I'm not gonna do that one myself)

**Comment by SkySkimmer** [comment]
Still waiting for fcl

**Comment by JasonGross** [comment]
> Ping @JasonGross (fiat crypto legacy is allow failure, I made almost all other overlays but I'm not gonna do that one myself)

I don't know why f-c-l is still allow failure.  It was originally put there due to the intricacy of the tactic scripts, but it has not had an issue that required patching the tactics in a long time, it no longer tries to maintain backwards compatibily, and it has caught bugs that no other CI development has caught.  I think it should be moved out of allowed failure.

I'll try to create an overlay within the next couple days to a week.

The fix is just to replace `:>` with `::` in classes?

File: doc/changelog/02-specification-language/18590-cleanup_16230.rst [hunk 2]
```diff
@@ -0,0 +1,9 @@
+- **Changed:**
+  warnings `future-coercion-class-constructor`
+  and `future-coercion-class-field` about ``:>`` in :cmd:`Class` as
+  error by default. This offers a last opportunity to replace ``:>``
+  with ``::`` (available since Coq 8.18) to declare typeclass instances
+  before making ``:>`` consistently declare coercions in all records in
+  next version
```
Comment (JasonGross) [inline doc/changelog/02-specification-language/18590-cleanup_16230.rst:7]: I wrote a script to do this replacement in f-c-l (PR/overlay at https://github.com/mit-plv/fiat-crypto/pull/1827), maybe it's worth linking to https://gist.github.com/JasonGross/59fc3c03664f2280849abf50b531be42 here?

Comment (proux01) [reply doc/changelog/02-specification-language/18590-cleanup_16230.rst]: @JasonGross thanks for the overlay! And thanks for the script, I added it, along with the one I used for iris, to the changelog.

**Comment by proux01** [comment]
@JasonGross fiat_crypto_legacy is still failing, apparently your script did not handle definitional typeclasses

**Comment by JasonGross** [comment]
> apparently your script did not handle definitional typeclasses

I fixed my script and made https://github.com/mit-plv/fiat-crypto/pull/1830, which will auto-merge once CI finishes running there

**Comment by proux01** [comment]
Thanks that seems to work

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**